controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java

```diff
@@ -23,26 +23,36 @@
 package org.jboss.as.controller;
 
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.locks.AbstractQueuedSynchronizer;
 
 /**
  * Basic lock implementation using a permit object to allow reentrancy. The lock will only be released when all
  * participants which previously acquired the lock have called {@linkplain #unlock}.
+ * The lock supports two mutually exclusive modes, shared and exclusive. If any shared locks are acquired and held
+ * then the exclusive lock may not be acquired, and if the exclusive lock is held, the shared locks may not be acquired.
+ * For an existing "permit holder" (operationId), the lock may be reentrantly re-acquired.
  *
  * @author Emanuel Muckenhuber
+ * @author Ken Wills
  */
 class ModelControllerLock {
     private final Sync sync = new Sync();
 
-    void lock(Integer permit) {
+    void lock(final Integer permit) {
         if (permit == null) {
             throw new IllegalArgumentException();
         }
         sync.acquire(permit);
     }
 
-    boolean lock(Integer permit, long timeout, TimeUnit unit) throws InterruptedException {
+    void lockShared(final Integer permit) {
+        if (permit == null) {
+            throw new IllegalArgumentException();
+        }
+        sync.acquireShared(permit);
+    }
+
+    boolean lock(final Integer permit, final long timeout, final TimeUnit unit) throws InterruptedException {
```

Comment:
This doesn't need to declare "throws InterruptedException" since it doesn't throw that.

This is a preexisting problem but the new "lockShared" has it too now, so might as well clean up the signature of both of them.